eventuals/collect.h

```diff
@@ -5,45 +5,70 @@
 #include "eventuals/loop.h"
 #include "eventuals/type-traits.h"
 
-// NOTE: The following include is a rather large one.
-#include "google/protobuf/repeated_field.h"
-
 ////////////////////////////////////////////////////////////////////////
 
 namespace eventuals {
 
 ////////////////////////////////////////////////////////////////////////
 
-template <typename Container>
+template <typename Collection, typename Enable = void>
+struct EventualsCollector {
+  template <typename T>
+  void Collect(Collection& collection, T&& value) {
+    static_assert(
+        !std::is_same_v<void, Enable>,
+        "Provide your own EventualsCollector");
+  }
+};
+
+////////////////////////////////////////////////////////////////////////
+
+// Collectors for STL collections.
+
+template <typename Collection>
+struct EventualsCollector<
+    Collection,
+    std::enable_if_t<
+        eventuals::HasEmplaceBack<Collection>::value>> {
+  template <typename T>
+  std::enable_if_t<
+      std::is_convertible_v<
+          T,
+          typename Collection::value_type>>
+  Collect(Collection& collection, T&& value) {
+    collection.emplace_back(std::forward<T>(value));
+  }
+};
+
+template <typename Collection>
+struct EventualsCollector<
+    Collection,
+    std::enable_if_t<
+        eventuals::HasInsert<Collection>::value>> {
+  template <typename T>
+  std::enable_if_t<
+      std::is_convertible_v<
+          T,
+          typename Collection::value_type>>
+  Collect(Collection& collection, T&& value) {
+    collection.insert(std::forward<T>(value));
+  }
+};
+
+////////////////////////////////////////////////////////////////////////
+
+template <typename Collection>
 [[nodiscard]] auto Collect() {
-  return Loop<Container>()
-      .context(Container())
-      .body([](auto& data, auto& stream, auto&& value) {
-        if constexpr (
-            std::is_same_v<
-                Container,
-                google::protobuf::RepeatedField<
-                    typename Container::value_type>>) {
-          data.Add(value);
-        } else if constexpr (
-            std::is_same_v<
-                Container,
-                google::protobuf::RepeatedPtrField<
-                    typename Container::value_type>>) {
-          // Should probably use std::forward instead of std::move,
-          // but using std::forward results in a compiler error,
-          // because google::protobuf::RepeatedPtrField::Add
-          // expects an rvalue reference.
-          data.Add(std::move<decltype(value)>(value));
-        } else if constexpr (HasEmplaceBack<Container>::value) {
-          data.emplace_back(std::forward<decltype(value)>(value));
-        } else {
-          data.insert(data.cend(), std::forward<decltype(value)>(value));
-        }
+  return Loop<Collection>()
+      .context(Collection())
+      .body([](auto& collection, auto& stream, auto&& value) {
+        EventualsCollector<Collection>().Collect(
```

Comment:
Let's make `Collect()` be a static function and then we can do:
```suggestion
        EventualsCollector<Collection>::Collect(
```